gui.py

```
@@ -65,6 +78,43 @@ def switch_to_download():
+    project_selectall_button.config(text="Select all", command=select_all_projects)
+
+# Login validation
+def validate_login():
+    success = download_controller.validate_login(user_entry.get(), pw_entry.get())
+    if not success:
+        messagebox.showerror("Login Failed", "Try again. Try not to mess up many times or Scratch might flag you as a clanker.")
+        return
+    switch_to_project_select()
+
+# Get project list everytime filter is reselected and show in the checklist
+def get_project_list(filter):
```

Ruff (0.15.7)

[error] Function argument `filter` is shadowing a Python builtin

(A002)